### PR TITLE
snap

### PR DESCRIPTION
**Pull request by mikeroyal**
Includes  snapcraft.yaml to build [snap](https://snapcraft.io/).

### Description

Provides a yaml file to package the CLI for Amazon EKS  as a snap. Snap makes the setup and install process easier for the end user by auto-updating to new versions or rollback to previous version if the newest version breaks.
### Checklist

- [ ] Code compiles correctly (i.e `make build`)
- [ ] Added tests that cover your change (if possible)
- [ ] All unit tests passing (i.e. `make test`)
- [ ] All integration tests passing (i.e. `make integration-test`)
- [ ] Added/modified documentation as required (such as the `README.md`, and `examples` directory)
- [ ] Added yourself to the `humans.txt` file

### CODE REVIEW

File: build/snap/snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+name: eksctl
+summary: CLI tool.
+description:| a CLI for Amazon EKS
+  
+adopt-info:eksctl
```
Comment (errordeveloper) [inline build/snap/snapcraft.yaml]:
```suggestion
adopt-info: eksctl
```

File: build/snap/snapcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
+name: eksctl
+summary: CLI tool.
+description:| a CLI for Amazon EKS
```
Comment (errordeveloper) [inline build/snap/snapcraft.yaml]:
```suggestion
description: a CLI for Amazon EKS
```

File: build/snap/snapcraft.yaml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    command: bin/eksctl
+    plugs:
+      - home
+      - network
+      - docker
+      - removable-media
+
+parts:
+  eksctl:
+  plugin: nil
```
Comment (errordeveloper) [inline build/snap/snapcraft.yaml]: Is this needed?

File: build/snap/snapcraft.yaml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  eksctl:
+    command: bin/eksctl
+    plugs:
+      - home
+      - network
+      - docker
+      - removable-media
+
+parts:
+  eksctl:
```
Comment (errordeveloper) [inline build/snap/snapcraft.yaml]: It doesn't look like valid YAML to me, indentation of the object below this this is off by one level.

File: build/snap/snapcraft.yaml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      last_committed_tag_ver="$(echo ${last_committed_tag} | sed 's/v//')"
+      last_released_tag="$(snap info $SNAPCRAFT_PROJECT_NAME | awk '$1 == "beta:" { print $2 }')"
+      # If the latest tag from the upstream project has not been released to
+      # beta, build that tag instead of master.
+      if [ "${last_committed_tag_ver}" != "${last_released_tag}" ]; then
+        git fetch
+        git checkout "${last_committed_tag}"
+      fi
+      snapcraftctl set-version "$(git describe --tags | sed 's/v//')"
+      override-build: |
```
Comment (errordeveloper) [inline build/snap/snapcraft.yaml]: Indentation seems to be off by here also, but the other way.

**Comment by errordeveloper** [comment]
Also, we will need to integrate this with goreleaser, so it becomes part of the release process - see docs here: https://goreleaser.com/customization/#Snapcraft.

**Comment by mikeroyal** [comment]
Hi, @errordeveloper  and @dholbach I will make the changes.

**Comment by errordeveloper** [comment]
@mikeroyal thanks!

**Comment by errordeveloper** [comment]
@mikeroyal how are you getting on? If you need help, please feel free to ask on [Slack](https://slack.weave.works).

File: build/snap/snapcraft.yaml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    command: bin/eksctl
+    plugs:
+      - home
+      - network
+      - docker
+      - removable-media
+
+parts:
+  eksctl:
+  plugin: nil
```
Comment (dholbach) [inline build/snap/snapcraft.yaml]:
```suggestion
    plugin: nil
```

File: build/snap/snapcraft.yaml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+      - removable-media
+
+parts:
+  eksctl:
+  plugin: nil 
+    source: https://github.com/weaveworks/eksctl.git
+    source-type: git
+    override-pull: |
+      git clone https://github.com/weaveworks/eksctl.git src/github.com/weaveworks/eksctl
+       cd src/github.com/weaveworks/eksctl
```
Comment (dholbach) [inline build/snap/snapcraft.yaml]:
```suggestion
      cd src/github.com/weaveworks/eksctl
```

File: build/snap/snapcraft.yaml [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      last_committed_tag_ver="$(echo ${last_committed_tag} | sed 's/v//')"
+      last_released_tag="$(snap info $SNAPCRAFT_PROJECT_NAME | awk '$1 == "beta:" { print $2 }')"
+      # If the latest tag from the upstream project has not been released to
+      # beta, build that tag instead of master.
+      if [ "${last_committed_tag_ver}" != "${last_released_tag}" ]; then
+        git fetch
+        git checkout "${last_committed_tag}"
+      fi
+      snapcraftctl set-version "$(git describe --tags | sed 's/v//')"
+      override-build: |
```
Comment (dholbach) [inline build/snap/snapcraft.yaml]:
```suggestion
    override-build: |
```

File: build/snap/snapcraft.yaml [hunk 9]
```diff
@@ -0,0 +1,55 @@
+name: eksctl
+summary: CLI tool.
+description: a CLI for Amazon EKS
+  
+adopt-info: eksctl
+
+grade: stable
+confinement: strict
```
Comment (dholbach) [inline build/snap/snapcraft.yaml]:
```suggestion
confinement: strict
base: core18
```

**Comment by dholbach** [comment]
I added a couple of suggestions to the `yaml`. It'd be great if you could accept them to at least get the yaml to validate.

I'd also recommend to take a look at the https://github.com/weaveworks/flux/pull/1966 it might simplify a few things for you. Also I don't get the snap to build.

Could you use the `make` plugin of `snapcraft` to just run `make build` as suggested in `CONTRIBUTING.md`? @errordeveloper @martina-if Why is there no `make install` step? That'd be ordinary "make" behaviour (and what the plugin expects - maybe we can work around that).

**Comment by dholbach** [comment]
I just checked, using `plugin: go` (and following some other its as described in the Flux PR from above) will make the build pass.

File: build/snap/snapcraft.yaml [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      - docker
+      - removable-media
+
+parts:
+  eksctl:
+    plugin: nil
+    source: https://github.com/weaveworks/eksctl.git
+    source-type: git
+    override-pull: |
+      git clone https://github.com/weaveworks/eksctl.git src/github.com/weaveworks/eksctl
```
Comment (errordeveloper) [inline build/snap/snapcraft.yaml]: Does snapcraft run builds for us, or do we have an option to upload? I'd expect with goreleaser we'd be uploading the same binaries we upload to GitHub, which is preferred.

**Comment by errordeveloper** [comment]
@dholbach does `plugin: go` use CGO? We should disable it, and use `release` and `netgo` tags. Ideally, as I said earlier, we should just use the binaries we build on release that are getting uploaded to GitHub.

**Comment by errordeveloper** [comment]
> Why is there no `make install` step?

It never works, one of worst OSS conventions I know of :)

**Comment by dholbach** [comment]
There's https://snapcraft.io/build which helps with the build.

**Comment by errordeveloper** [comment]
> There's https://snapcraft.io/build which helps with the build.

That's great, but I'd rather avoid having to define the release process in multiple places and/or in multiple different ways. Can we please look at how it works with goreleaser?

**Comment by mikeroyal** [comment]
Hi @errordeveloper, the go plugin uses the go get package installer.

**Comment by dholbach** [comment]
I just pushed https://github.com/mikeroyal/eksctl/pull/1 which makes the build work fine for me. Please let me know what you think, @mikeroyal !

**Comment by mikeroyal** [comment]
Looks good to me @dholbach

**Comment by dholbach** [comment]
To tie this up, I pushed https://github.com/weaveworks/eksctl/pull/969 with all your commits @mikeroyal - thanks again for your work on this!

**Comment by errordeveloper** [comment]
Closing, as commits are no included in #696.

**Pull request closed without merging**